File: mixin-padang-garut/src/main/python/garut/functions/dataset/split_column.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
import pandas as pd
from typing import Dict, List
from garut.contexts.context import Context
from garut.functions.splitter.split_by_none import SplitByNone
from garut.functions.splitter.splitter import Splitter
from garut.instruction import InstructionFactory
from garut.functions.function import FunctionRegistry
from garut.functions.dataset.dataset_function import DatasetFunction


class SplitColumn(DatasetFunction):

    FUNCTION_NAME = "SplitColumn"

    COLUMN = "column"
    SPLITTER = "splitter"

    def execute(self, context: Context, options: Dict[str, any]) -> any:

        # Options
        dataset = self.getDataset(options)
        column = options.get(SplitColumn.COLUMN, None)
        splitter: Splitter = options.get(SplitColumn.SPLITTER, SplitByNone())

        # Initial
        columns: List = dataset.columns.values.tolist()
        index = columns.index(column)
        result = dataset.copy()

        # Split
        split: pd.DataFrame = splitter.split(result[column])
        newColumns = split.columns.values.tolist()
        result = pd.concat([result, split], axis=1)
        result.drop(columns=[column], inplace=True)
        columns.remove(column)

        # Position
        columns[index:index] = newColumns
        return result[columns]


parameters = [DatasetFunction.DATASET, SplitColumn.COLUMN, SplitColumn.SPLITTER]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(SplitColumn.FUNCTION_NAME, SplitColumn, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(SplitColumn.FUNCTION_NAME)

File: mixin-padang-webface/src/main/js/padang/functions/dataset/SplitColumn.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */
import FunctionPlan from "padang/plan/FunctionPlan";
import ParameterPlan from "padang/plan/ParameterPlan";
import ParameterPlanUtils from "padang/plan/ParameterPlanUtils";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import ColumnFunction from "padang/functions/dataset/ColumnFunction";
import DatasetFunction from "padang/functions/dataset/DatasetFunction";

import SplitByDelimiter from "padang/functions/splitter/SplitByDelimiter";
import SplitByPositions from "padang/functions/splitter/SplitByPositions";

export default class SplitColumn extends DatasetFunction {

	public static FUNCTION_NAME = "SplitColumn";

	constructor(
		public dataset: string,
		public column: string,
		public splitter: any) {
		super(SplitColumn.FUNCTION_NAME, dataset);
	}

	public static COLUMN_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"column",
		"Column",
		"Column to be splitted",
		"",
		ColumnFunction.getDefaultColumnKeysAssignable()
	);
	public static SPLITTER_PLAN: ParameterPlan = ParameterPlanUtils.createAnyPlan(
		"splitter",
		"Splitter",
		"Splitter function",
		"=[" +
		SplitByDelimiter.FUNCTION_NAME + ", " +
		SplitByPositions.FUNCTION_NAME +
		"]",
		SplitByDelimiter.FUNCTION_NAME + "(',', false, -1)"
	);

	public static getPlan(): FunctionPlan {
		let plan = new FunctionPlan(
			SplitColumn.FUNCTION_NAME,
			"Split Column",
			"mdi-arrow-split-vertical",
			"Split a string column into multiple columns"
		);
		let parameters = plan.getParameterList();
		parameters.add(DatasetFunction.DATASET_PLAN);
		parameters.add(SplitColumn.COLUMN_PLAN);
		parameters.add(SplitColumn.SPLITTER_PLAN);
		return plan;
	}

}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(SplitColumn.getPlan());